New tests and renaming `solve` to `iterate`

pwl opened the pull request:
Among other fixes, I removed the `tdir` from dense output, it wasn't necessary there and it didn't work with my custom integrator example.  I think `tdir` should be internal to the integrator implementation and not a part of a general backend API.

I also included a prototype of a test function that ensures that the minimal interface is met and that a solver works for compatible problems, for the example usage refer to `examples/custom_integrator.jl`.  I found some bugs in `EulerIntegrator` this way, so it could be useful for other users.  And the problem set should be expanded if we like this design.  Because of the possibly large amount of test cases, and the fact that this is not really a core part of `ODE.jl` we could move the tests to a separate module (`ODETests`?).

pwl committed:
More fixes, removed `tdir` from dense output

pwl committed:
Fixes for the custom integrator

pwl committed:
Prototype of a test function for custom integrators

pwl committed:
Added an iterator trait to Problem

mauro3 commented on src/test_solver.jl:
```diff
@@ -0,0 +1,63 @@
+import Base.Test: @test, @test_approx_eq_eps
```
Name the file `test_integrator.jl` for consistency.

This overlaps with `test/iterators.jl`, maybe refactor to have only one?

pwl replied:
How about `src/odetests/iterators.jl`? Or `src/odetests/test_iterators.jl`?

mauro3 commented on src/test_solver.jl:
```diff
@@ -0,0 +1,63 @@
+    # filter the tests
+    tests = filter(x->in(typeof(x[:ivp]),eqntypes),testset)
+
+    for test in testset
```
Would it make sense to make each sub-test a function?  That would make documenting the test easier and would allow adding more tests without making this function huge?

pwl replied:
Like `test_harmonic_oscillator`?

mauro3 commented on src/test_solver.jl:
```diff
@@ -0,0 +1,63 @@
+import Base.Test: @test, @test_approx_eq_eps
+
+const testset = [
```
As suggested in #20 and #21, maybe have the ode and y0 separate to mix and match odes and types.  But this can also be done in a later commit/PR.

pwl replied:
This is totally just a simple prototype.  This would be a cool idea, on the other hand, do we actually need that many different ODEs?  Is there anything that makes each of them independent enough to justify the presence of all of them?  If a solver can solve a harmonic oscillator to a reasonable precision then it surely will be able to solve other non-stiff equations.  Although, keeping additional ODEs doesn't seem to have any disadvantages to me: they don't impact the performance of the test suite that much.

mauro3 commented on src/base.jl:
```diff
@@ -176,6 +176,21 @@ Base.length(prob::Problem) = length(prob.solver)
+        errgen=(y-[exp(t)] for (t,y) in sol)
+        errout=collect(errgen)
+
+    TODO: doesn't work for 0.4 and might have show issues due to non-copying output
```
I'd be fine with dropping 0.4 compatibility.  But then maybe we shouldn't as other packages will take some time to adapt 0.5.

pwl replied:
I wouldn't mind dropping 0.4, then we could also drop the dependency on `Compat`.

mauro3 commented on src/dense.jl:
```diff
@@ -48,12 +48,28 @@ end
+                                               tout::AbstractVector{T} = T[tstop],
                                                opts...)
+    if all(tout.>=ivp.t0)
+        tout = sort(tout)
```
We should error on un-sorted and not sort it.

pwl replied:
That's another way to handle it, but then which convention do we choose with `t0=0`?
- `tout=[0,-1/2,-1]`
- `tout=[-1,-1/2,0]`

pwl replied:
With `sort` we don't have to restrict ourselves to any of these, although now that I think about it, if we sort `tout` we loose the 1:1 relationship with the order of `tout` the user provided.

mauro3 replied:
I think most of the time passing in a unsorted `tout` is a mistake on the side of the user.  If he/she indeeds wants unsorted, then it is trivial to unsort afterwards.

pwl replied:
So do we choose `tout=[-1,-1/2,0]` with `t0=0`?

mauro3 replied:
I don't understand you concern, maybe you need to explain more carefully.  My point was that `tout` needs to be sorted in `tdir` direction, if it's not then error.

mauro3 commented on src/dense.jl:
```diff
@@ -48,12 +48,28 @@ end
+
 @compat function (::Type{DenseOutput{I}}){T,I}(ivp::IVP{T};
+                                               tstop = T(Inf),
+                                               tout::AbstractVector{T} = T[tstop],
```
Maybe we should make `tout` a non-optional argument. What is the point of calling a dense solver to only output `tstop`?

pwl replied:
According to our convention for solver constructors we only allow for `DenseOutput(ivp;opts...)`, so we can't add/remove nonoptional arguments.

mauro3 commented on src/dense.jl:
```diff
@@ -48,12 +48,28 @@ end
 
 Base.length(dense::DenseOutput) = length(dense.opts.tout)
 
+tdir(ivp, solver::DenseOutput) = sign(solver.opts.tstop - ivp.t0)
```
What is this needed for?

pwl replied:
nothing now

mauro3 commented:
I like the test.  I think it should be run for the "internal" solver too, replacing `test/iterators.jl`.   As you suggested in #20 and #21, the tests-sets could be broken up into the the equation, `y0` and integrators.  Those three could then be mixed to make the full test case.  But we can merge like this for now and resolve that later.

How did the `tdir` not work?  Note, doing `t2,t1=t1,t2`, is taking time-direction into account.  If that is necessary, then I'd vote for doing it the same way as elsewhere.

pwl commented:
I removed `tdir` because it is not a part of our API (we don't demand that solvers implement it).  Indeed, dense output wasn't working for the example of an integrator I produced.  We could add `tdir` to the API but this is not strictly necessary for dense output, as demonstrated in this patch.

pwl committed:
Minor doc update

pwl committed:
More minor fixes

pwl committed:
New tests module

pwl committed:
New test suite, including custom types

pwl committed:
Even better tests (via @testset)

pwl committed:
New naming convention and Solution type

pwl committed:
Moved the ODETest to a submodule

pwl committed:
Removed an old `tdir` from dense output

pwl commented:
I updated a whole bunch of stuff

- `solve` now solves the equation and `iterate` generates an iterator (this fixes #17)
- New return type, `Solution`, returned by `solve`
- I completely re-wrote the tests, we are now using test sets and we have a single function to run if you want to test a user defined iterator.  There is also a test case with custom (minimal) types for `Y` and `T` (fixes #20).
- We now have a working example of user defined iterator with the test suite.  Although tests fail for some reason, probably some simple bug in the algorithm but I can't find it.

TODO before merging to master:
- update the docs to reflect the renaming of `solve` to `iterate` and add some info on `Solution`
- API for `Solution`, what do we want from it?  Surely we have to be able to extract `t`, `y` and `dy`, but how do we extract them?  Can we leave it as is for now?

pwl committed:
Updated docs

pwl commented:
Docs are updated, I think the API for `Solution` can wait for the next iteration.  How about we merge this?

mauro3 commented:
I'll try to look at it tonight. If not then merge it tomorrow.

mauro3 commented on docs/src/index.md:
````diff
@@ -29,7 +29,52 @@ F(t,y) = y
+
+```@example solve
+sol = ODE.solve(ode, tstop = 1)
+```
````
Whilst a default tstop=Inf is reasonable for an iterator, maybe less so here.  How about it is a mandatory argument?

pwl replied:
Totally agree, we could make it the required argument.  I actually run into this exact problem myself:-)

pwl replied:
Alternatively we could check if `tstop` is finite and spit out an error if it isn't.  On top of that, remember, that if we implement `stopcondition` or whatever you call it `tstop` won't be necessary to get a solution in finite steps.  I think we could actually leave it as is.

pwl replied:
One more solution would be to have `maxiters` hard coded into the iterator and set to some reasonable finite number by default.  This would prevent any runaway integrations.

mauro3 replied:
No, those are all round-about ways.  The error should be that someone forgot to specify `tstop`. Using the defaults and then an error is thrown is odd, or that iteration stops after maxiter, which may be any amount of time depending on the equation.

pwl replied:
How about an error if ``!isfinite(tstop)` and if `stopcondition` is undefined (but only for `solve`)?  For now, because `stopcondition` is unimplemented, we would throw an error on `!isfinite(tstop)`.  Leaving it as a keyword argument would make it look clearer and more consistent with `iterate`.

mauro3 replied:
We can leave it as is for now.  I don't think `iterate` and `solve` need to have the same signatures though as they do quite different things.

pwl replied:
It looks weird to me to have a function mixing high level arguments like `IVP` with simple numbers in a single call: `solve(ivp, 1)`.  It is way clearer what the number means if we write it as `solve(ivp,tstop=1)` and error on `!isfinite`.  It also leaves us more flexibility in how to treat the infinite `tstop` case.  And as I already mentioned, strictly speaking `tstop` is not the only way to limit the integration.

mauro3 replied:
Ok

mauro3 commented on docs/src/index.md:
````diff
@@ -39,68 +84,85 @@ iterator interface.
-for (t,y) in sol
+```@example iterate
+iter = ODE.iterate(ode)
+for (t,y) in iter
````
`for (t,y) in ODE.iterate(ode)` seems clear enough.

mauro3 commented on docs/src/index.md:
````diff
@@ -39,68 +84,85 @@ iterator interface.
+
+```@example iterate
+iter = ODE.iterate(ode; tstop = 1)
+for (t,y,dy) in iter
````
Actually, is not the more common case to know tstop?  Thus make it a mandatory argument here too?  The ones who want endless iteration can just supply Inf.

pwl replied:
I would rather leave the `tstop` as an optional argument here.  In many cases you continue the integration until you hit a specific condition and you don't know when this might happen.  It would feel awkward to specify the `tstop` if you don't care about it.

mauro3 commented on docs/src/index.md:
```diff
@@ -39,68 +84,85 @@ iterator interface.
+   we don't know a priori how many steps the integration will require
+   (especially in the case of adaptive solvers).  This means that the
+   functions requireing `length` might not work.  For the same reason
+   there are no `getindex` methods.
```
Aside: `getindex` should also have ~O(1) complexity.

mauro3 commented on src/base.jl:
```diff
@@ -7,11 +7,25 @@
+"""
+    AbstractIVP{T,Y}
+
+A progenitor of types representing an IVP (initial value
```
Had to look `progenitor` up, maybe something less fancy? "The abstract supertype of all IVP..."

pwl replied:
I have no idea how I came up with "progenitor", I will change it.

mauro3 commented on src/base.jl:
```diff
@@ -7,11 +7,25 @@
 
 """
 
+    IVP{T,Y,F,G,J} <: AbstractIVP{T,Y}
```
There is this: https://github.com/JuliaDocs/DocStringExtensions.jl

I think it can insert those signatures directly.

pwl replied:
I will take a look at it.

pwl replied:
I looked it up and tested it.  It seems it's not working properly in all cases and the produced output is not very clear (e.g. we would like to highlight some optional arguments and their defaults).  I would leave it as is for now.

mauro3 replied:
Thanks! (If you got time, file issues with DovStringExtentions.jl, as having this automatic would be much better.)

mauro3 commented:
Yep, looks good modulus a few minor comments.  Thanks for the good work!

One thing though: change the title to reflect what went on, in case we ever need to dig it up.

pwl commented:
By the way, should we drop 0.4 compatibility now?

mauro3 commented:
I'm not quite sure anymore about dropping 0.4; how long will it take for packages and users to move to 0.5?  But for sure we should not optimize for it.  Actually, this is pre 1.0 Julia, so let's just drop 0.4.

pwl committed:
Fixed infinite tstop bug custom type test

pwl commented:
I dropped the 0.4 compatibility and added the errors for infinite `tstop`, they should be able to catch most of the problematic use cases.  I also improved the implementation of the minimal type: now it's impossible to construct the custom float type from `Float64`, which will allow us to catch most of the precision bugs within new integrators.  This test might seem to be too strict in the beginning but it should act as a go to for new integrators (you can always opt-out from it by removing it from the list of test cases).

pwl commented:
Should we remove `test/interface-tests.jl` now that we have `src/tests/minimal_types.jl`?

pwl committed:
Better handling of function types in IVP

pwl committed:
Comment cleanup

pwl committed:
Switched to the Documenter HTML generation

pwl committed:
Wrapped all the tests into testsets

pwl committed:
Removed interface-tests.jl

pwl commented:
How about we merge it now?
